### ISSUE TITLE
Bug reports

### ISSUE THREAD

**Issue by p-morais** [deepmind/mujoco#13]
Hi,

First off, thank you for taking on this effort to open source and maintain MuJoCo, it's really greatly appreciated. I've been using MuJoCo for a while now so figure it may be nice to formally report some bugs I've noticed here. If you prefer I can split each of these into individual bug reports/feature requests.

- Cylinder-box contacts are extremely noisy. I believe Emo was aware of this deficiency and planning to fix it in a future release. Is that still the case?
- mj_contactForce sometimes returns what looks like garbage data, usually only once at the initiation of contact. This seems like it's perhaps an uninitialized memory bug. I can provide a minimal reproducible example at some point if that helps. 
- Contact distance calculations don't scale properly with certain geom pairs. I can't remember the details of this bug, but I reported it in the MuJoCo forum once so once that's up and running again I can update this. I don't think this affects normal simulation, but shows up if you set margin=gap>>0

Finally, at some point Emo said he was working on an incremental compiler for MuJoCo. Is that also on the development roadmap still?

**Comment by yuvaltassa** [comment]
Hi @p-morais !

We'd still love to see a MWE for your second and third items.

Regarding the first item, I recommend you take a look at the `multiCCD` option added in [MuJoCo 2.1.5](https://mujoco.readthedocs.io/en/latest/changelog.html#version-2-2-0-may-23-2022). In the meantime I'm closing this issue.

**Comment by milutter** [comment]
Hi @saran-t & @yuvaltassa 

I came across the same issue with the `mj_contactforce`. I am able to provide a minimal example using the `dm_control` `hoppper`.  An interesting observation is that the nearly infinite contact forces (1e+281) only appear if I re-instantiate the physics object multiple times (which might hint towards a memory issue). If the physics object is only created once, the contact forces are well-behaved. The frequency of the nearly infinite forces is stochastic (despite seeding numpy). 

The minimal example below creates a `dm_control.mujoco.Physics` instance with the default `hopper.xml`, randomly initializes the joint angles and steps the simulation. After each simulation step, the contact force is computed using `physics.data.contact_force(int)`. If the contact force is greater than `1e+12`, the script prints these forces and other relevant information from `physics.data`. Below you can find the code and the example output. MuJoCo and DM_control are the latest version installed via pip (`MuJoCo Version 2.3.0`, `DM Control Version = 1.0.8`)

**Example Output:**
```
###########################
Without Reloading the Physics Object


###########################
With Reloading the Physics Object

#### Episode = 3 - Step = 83
Contact Force = [+1.149e+257, +5.083e+70, +1.149e+257, +0.000e+00, +0.000e+00, +0.000e+00,]
Contact Distance = -0.013621119765079931
EFC Force = [+5.083e+70, -9.033e-167, +1.149e+257, +5.083e+70, -9.033e-167, +1.149e+257, +5.083e+70, -9.033e-167,]
EFC B = [-1.378e-171, +1.149e+257, +4.910e+70, -1.378e-171, +1.149e+257, +4.910e+70, -9.033e-167, +1.149e+257,]

Contact Force = [+1.149e+257, -1.149e+257, +5.083e+70, +0.000e+00, +0.000e+00, +0.000e+00,]
Contact Distance = -0.006118167385443178
EFC Force = [+5.083e+70, -9.033e-167, +1.149e+257, +5.083e+70, -9.033e-167, +1.149e+257, +5.083e+70, -9.033e-167,]
EFC B = [-1.378e-171, +1.149e+257, +4.910e+70, -1.378e-171, +1.149e+257, +4.910e+70, -9.033e-167, +1.149e+257,]


#### Episode = 6 - Step = 77
Contact Force = [+1.601e+281, +1.795e+85, +1.601e+281, +0.000e+00, +0.000e+00, +0.000e+00,]
Contact Distance = -0.012206859106728418
EFC Force = [+1.795e+85, -1.349e-142, +1.601e+281, +1.794e+85,]
EFC B = [+1.601e+281, +1.892e+85, -8.840e-138, +1.601e+281,]


#### Episode = 8 - Step = 44
Contact Force = [+2.500e+247, -2.500e+247, +1.022e+61, +0.000e+00, +0.000e+00, +0.000e+00,]
Contact Distance = -0.004460699324460478
EFC Force = [-4.435e-186, +2.500e+247, +1.022e+61, -4.435e-186,]
EFC B = [+1.022e+61, -4.435e-186, +2.500e+247, +1.022e+61,]
```


**Example Code:**
```
from importlib import metadata

import mujoco
import numpy as np
from dm_control import mujoco as dm_mujoco
from dm_control.suite import hopper
from dm_control.suite.utils import randomizers

np.set_printoptions(suppress=True, precision=5, linewidth=500,
                    formatter={"float_kind": lambda x: f"{x:+0.3e},"})

if __name__ == "__main__":

    print("\n\n\n")
    print(f"MuJoCo Version {mujoco.__version__}")
    print(f"DM Control Version = {metadata.version('dm_control')}\n\n")

    n_episodes = 100
    n_steps = 100

    def run_episodes(reload_physics):
        np.random.seed(0)
        random = np.random.RandomState(None)
        xml_model, assets = hopper.get_model_and_assets()

        for i in range(n_episodes):

            # Construct / Reload physics
            if i == 0 or reload_physics:
                physics = dm_mujoco.Physics.from_xml_string(xml_model, assets)

            # Initialize qpos randomly
            randomizers.randomize_limited_and_rotational_joints(physics, random)

            # Simulate and episode of n_steps
            done = False
            for step_i in range(n_steps):
                physics.step()

                for j in range(physics.data.ncon):
                    contact_force = physics.data.contact_force(j).flatten()

                    if np.max(contact_force) > 1.e12:
                        print(f"\n#### Episode = {i} - Step = {step_i}")
                        done = True

                        for k in range(physics.data.ncon):
                            print(f"Contact Force = {physics.data.contact_force(k).flatten()}")
                            print(f"Contact Distance = {physics.data.contact.dist[k]}")
                            print(f"EFC Force = {physics.data.efc_force}")
                            print(f"EFC B = {physics.data.efc_b}\n")

                        break

                if done:
                    break

    # Run without re-loading the physics object
    print("\n\n###########################")
    print("Without Reloading the Physics Object")
    run_episodes(reload_physics=False)

    # Run without re-loading the physics object
    print("\n\n###########################")
    print("With Reloading the Physics Object")
    run_episodes(reload_physics=True)
```

**Comment by yuvaltassa** [comment]
Thanks @milutter! 

We'll look into it asap.

**Comment by saran-t** [comment]
@milutter Could you please check for me if adding the following lines immediately after your `physics.step()` resolves the issue?

```mujoco.mj_fwdActuation(physics.model.ptr, physics.data.ptr)
      mujoco.mj_fwdAcceleration(physics.model.ptr, physics.data.ptr)
      mujoco.mj_fwdConstraint(physics.model.ptr, physics.data.ptr)```

**Comment by milutter** [comment]
@saran-t Adding these three lines after `physics.step()` solves the problem. To be more specific, in this example calling `mujoco.mj_fwdConstraint` is already sufficient to resolve the problem. 

Having to call the constraint solver manually after each step seems counterintuitive as I would assume that `mj_step` and `mj_step2` have already run the constraint solver?

**Comment by yuvaltassa** [comment]
@milutter allow me to explain.

tl;dr don't use dm_control's, `physics.step()`, call `mujoco.mj_step()` (either from `dm_control` or just using the native bindings (recommended)).

DM Control has a bad design flaw (once upon a time considered a feature :slightly_smiling_face: ) that's been around for years. It is described in page 6 of [our publication](https://arxiv.org/abs/2006.12983). `physics.step()` breaks the step into the position+velocity component `step1` and the acceleration+integration `step2`. It assumes that `step1` has already been called in `reset_context` and then a `step()` first calls `step2`, advancing the state, and then `step1`, computing the position and velocity dependent quantities. See code [here](https://github.com/deepmind/dm_control/blob/e3d4c0aecfd710cd7428eb77af203dc5fbdacc0f/dm_control/mujoco/engine.py#L147). The benefit of this approach is that position and velocity dependent observations are w.r.t the current state rather than the previous step. Acceleration/force-dependent sensors, like touch sensors, of course cannot be synchronised with the current state, since they depend on the controls.

The bug, which we unfortunately did not pick up on for a long time, involves anything related to contact forces (touch sensors or just reading contact forces, like you are doing). The issue is that the position and identity of contact frames is position dependent, while the forces themselves are force/acceleration dependent. So in dm_control, when you a reading a contact force after calling `physics.step()`, you are getting the force from the _previous step_, read in the frame of the contact at the _current step_. Sometimes, when the number of contacts increases between these steps, your contact maybe refer to a non-existent value in the constraint force vector, and you will get garbage.

There are 3 obvious ways to ameliorate this:
1. (as a user, right now) Don't use `physics.step()` but rather `mj_step()`.
2. (todo, dm_control authors) Add an option to `physics.step()` to use `mj_step()`.
3. (todo, dm_control authors) Make sure calls to `physics.data.contact_force()` always call the 3 lines sent by @saran-t above (wasteful, but will be correct).

I deeply apologise for this. Unfortunately fixing old things is not easy due to the dependencies, so e.g. just fixing `physics.step()` by default is not an option. We will hopefully implement option 2 and maybe option 3 in the near future.

**Comment by milutter** [comment]
Thanks @yuvaltassa for the explanation. I was aware of the different `mj_step1` & `mj_step2` variants but not about all the implications on the contact force computation. I know that this cannot be easily fixed due to legacy code. It would be great to add/link this explanation about the limitations to the `physics.step` implementation such that problem becomes clearer when looking at the code.
